Deprecation guide for jQuery removal (#255)
ember-learn/deprecation-app: content/ember/v3/jquery-apis.md

---
id: jquery-apis
title: Replace jQuery APIs
until: '4.0.0'
since: '3.8'
---

> As jQuery will be removed from Ember by default, APIs that are coupled to jQuery have been deprecated.

Reviewer: ```suggestion
As of Ember 3.2.0, Ember no longer requires that all applications include jQuery, therefore APIs that are coupled to jQuery have been deprecated.
```

Author: I believe making jQuery optional was implemented for Ember.js 3.3, wasn't it? 
But there was another thing, IIRC the 3.3 release of Ember CLI was postponed (making it 3.4), which included the necessary changes required by the (first) RFC. So we referred to 3.4 as the version which made jQuery optional, as this was the least confusing option for users (who usually have Ember & CLI in lockstep). This is at least what the Guides say: https://guides.emberjs.com/release/configuring-ember/optional-features/#toc_jquery-integration

So I changed this to 3.4.0 now, hope this is fine?

> The rationale behind it is that jQuery is not needed by Ember itself anymore, and as many apps (e.g. mobile apps) are 

Reviewer: ```suggestion
Since jQuery is not needed by Ember itself anymore, and many apps (e.g. mobile apps) are 
```

> sensitive about performance in general and here in particular their bundle size, it is often beneficial for those to

Reviewer: ```suggestion
sensitive about performance, it is often beneficial for those to
```

> not ship with jQuery. If however this is not a major concern for your app, and you see value in using jQuery, it is 

Reviewer: ```suggestion
avoid shipping jQuery. If this is not a major concern for your app, and you see value in using jQuery, it is 
```

> absolutely legit to continue doing so. It is just not included by default anymore, and you would have to opt in to

Reviewer: ```suggestion
absolutely fine to continue doing so. It is just not included by default anymore, so you have to opt in to
```

using it with the `@ember/jquery` package as described below.

For addons it is a bit different as they are not aware of the context in which they are used. Any addon that still 
relies on jQuery will either force their consuming apps to continue bundling jQuery, or will not be usable for apps

> that decide not to do so. Therefor it is highly recommended to not rely on jQuery in general, unless there is a good

Reviewer: ```suggestion
that decide not to do so. Therefore it is highly recommended to avoid relying on jQuery in general, unless there is a good
```

reason (e.g. an addon wrapping a jQuery plugin).

#### Added deprecations

> The main jQuery integration API that has been deprecated is the component's `$()` method, which would give you a 

Reviewer: ```suggestion
The main jQuery integration API that has been deprecated is `this.$()` inside of an `Ember.Component`, which would give you a 
```

> jQuery object of the component's element. So instead of that you can use the `element` property, which provides a 

Reviewer: ```suggestion
jQuery object of the component's element. Instead, you can use the `this.element` property, which provides a 
```

    this.$('.animated').on('transitionend', () => this.doSomething());
  }
});
```

should be changed to:

```js
import Component from '@ember/component';

export default Component.extend({
  waitForAnimation() {
    this.element.querySelectorAll('.animated')
      .forEach((el) => el.addEventListener('transitionend', () => this.doSomething()));
  }
});
```

This applies in a similar fashion to component tests using the `setupRenderingTest()` helper. Instead of using 
`this.$()` in a test, you should use `this.element` (or alternatively the `find()`/`findAll()` helpers from 
`@ember/test-helpers`):

```js
test('it disables the button', async function(assert) {
  // ...
  
  assert.ok(this.$('button').prop('disabled'), 'Button is disabled');
});
```

should be changed to:

```js
test('it disables the button', async function(assert) {
  // ...
  
  assert.ok(this.element.querySelector('button').disabled, 'Button is disabled');
});
```

If you do continue to use jQuery, please make sure to always import it like this:

```js
import jQuery from 'jquery';
```

Accessing it from the `Ember` namespace as `Ember.$` is and will remain deprecated.

#### Opting into jQuery

> For apps and addons which are ok to work only with jQuery, you can opt into the jQuery integration now provided by

Reviewer: ```suggestion
Apps and addons which require jQuery, can opt into the jQuery integration now provided by
```

> the `@ember/jquery` package. This will provide the `Ember.Component.$()` API without any deprecation warnings, and

Reviewer: ```suggestion
the `@ember/jquery` package. This will provide the `this.$()` API to `Ember.Component`s without deprecation warnings, and
```

---
id: jquery-apis

> title: Remove jQuery APIs

Reviewer 2: ```suggestion
title: Replace jQuery APIs
```

Author: I can do that, by why do you think it's better? As we don't introduce any new APIs as a replacement, we are actually only removing some.

Reviewer 2: I checked existing deprecations and the phrasing seems to generally be the action that you need to do. While we did not introduce new APIs, we suggest that people use `this.element` and DOM APIs because as of 3.0 they are present in browsers supported by Ember.
It's fine either way!

sensitive about performance in general and here in particular their bundle size, it is often beneficial for those to
not ship with jQuery. If however this is not a major concern for your app, and you see value in using jQuery, it is 
absolutely legit to continue doing so. It is just not included by default anymore, and you would have to opt in to
using it with the `@ember/jquery` package as described below.

For addons it is a bit different as they are not aware of the context in which they are used. Any addon that still 
relies on jQuery will either force their consuming apps to continue bundling jQuery, or will not be usable for apps
that decide not to do so. Therefor it is highly recommended to not rely on jQuery in general, unless there is a good
reason (e.g. an addon wrapping a jQuery plugin).

#### Added deprecations

The main jQuery integration API that has been deprecated is the component's `$()` method, which would give you a 
jQuery object of the component's element. So instead of that you can use the `element` property, which provides a 
reference to a native DOM element:

> ```js

Author: Done

this.$().on('transitionend', () => this.doSomething());
```

should be changed to:

```js
this.element.addEventListener('transitionend', () => this.doSomething());
```

If you used `this.$()` to query for child elements, you can do so as well with native DOM APIs:

```js
this.$('.animated').on('transitionend', () => this.doSomething());
```

should be changed to:

```js
this.element.querySelectorAll('.animated')
  .forEach((el) => el.addEventListener('transitionend', () => this.doSomething()));
```

This applies in a similar fashion to component tests using the `setupRenderingTest()` helper. Instead of using 
`this.$()` in a test, you should use `this.element` (or alternatively the `find()`/`findAll()` helpers from 
`@ember/test-helpers`):

```js
test('it disables the button', async function(assert) {

>   ...

Author: Done

  
  assert.ok(this.$('button').prop('disabled'), 'Button is disabled');
});
```

should be changed to:

```js
test('it disables the button', async function(assert) {
  ...
  
  assert.ok(this.element.querySelector('button').disabled, 'Button is disabled');
});
```

If you do continue to use jQuery, please make sure to always import it like this:

```js
import jQuery from 'jquery';
```

> Accessing it from the `Ember` namespace as `Ember.$` (either directly or through destructuring) is and will remain 

Author: Hm, not that I'm aware of, is it!?

deprecated.

#### Opting into jQuery

For apps and addons which are ok to work only with jQuery, you can opt into the jQuery integration now provided by
the `@ember/jquery` package. This will provide the `Ember.Component.$()` API without any deprecation warnings, and
will make sure that the `EventDispatcher` will provide jQuery events to a component's event handler methods to 
maintain compatibility.  

```bash
ember install @ember/jquery
ember install @ember/optional-features
ember feature:enable jquery-integration
```

> For addons make sure that `@ember/jquery` is added as a `dependency` in its `package.json`!

Reviewer 2: ```suggestion
For addons make sure that `@ember/jquery` is added as a `dependency` in its `package.json`!
```

Reviewer 2: I wonder what happened there 🤔
